lib/Sema/TypeCheckPattern.cpp

```diff
@@ -475,15 +400,18 @@ class ResolvePattern : public ASTVisitor<ResolvePattern,
   // Member syntax 'T.Element' forms a pattern if 'T' is an enum and the
   // member name is a member of the enum.
   Pattern *visitUnresolvedDotExpr(UnresolvedDotExpr *ude) {
-    SmallVector<IdentTypeRepr *, 2> components;
-    if (!ExprToDeclRefTypeRepr(components, Context).visit(ude->getBase()))
+    // In synthesized code this will have been a pattern from the get-go.
+    if (ude->isImplicit())
+      return nullptr;
+
+    auto *repr = swift::parseSimpleTypeInRange(
+        DC->getParentSourceFile(), ude->getBase()->getSourceRange());
```

Comment:
I’m not a huge fan of re-parsing the type. Why can’t do an approach similar to `ExprToDeclRefTypeRepr` that rewrites the `UnresolvedDotExpr` into a `TypeRepr`?

Reply:
I thought it more prudent to reuse an existing transform than writing up a new one, which would add to maintenance and, possibly, bugs, with all the different nodes and representational quirks of expressions. Re-parsing is less efficient, on the other hand. 

___

Interestingly, there is a batch code completion test crashing along this code path because the buffer IDs of the `SourceLoc` and the context `SourceFile` in `TypeCheckASTNodeAtLocRequest` don’t match. Is this expected?

Reply:
> I thought it more prudent to reuse an existing transform than writing up a new one, which would add to maintenance and, possibly, bugs, with all the different nodes and representational quirks of expressions. Re-parsing is less efficient, on the other hand.

CC @rintaro who probably has opinions

> Interestingly, there is a batch code completion test crashing along this code path because the buffer IDs of the SourceLoc and the context SourceFile in TypeCheckASTNodeAtLocRequest don’t match. Is this expected?

Yes, that’s expected. When we are reusing the ASTContext between multiple completions, we are just re-parsing a function body and that function body will live in a separate buffer than the rest of the source file because there might have been modifications to that function body since the last completion.

Reply:
So.. I have come to realize that re-parsing is a bad idea. The only thing that matters right now when resolving an enum element pattern is whether the expression resolves to enum element with a type qualifier, but that’s not enough, because an identifier might be something else that shadows a type. This compiles:

```swift
enum E { case x }
struct S {
  let x = E.x
}
func foo(e: E) {
  let E = S()
  switch e {
    case E.x: break
  }
}
```
...but should it?

cc @xedin